apps/hash-ai-worker-ts/src/activities/shared/get-llm-response.ts

```diff
@@ -0,0 +1,467 @@
+    }
+
+    if (retryMessageContent.length > 0) {
+      return retry({
+        retryMessage: {
+          role: "user",
+          content: retryMessageContent,
+        },
+      });
+    }
+  }
+
+  const response: LlmResponse<AnthropicLlmParams> = {
+    status: "ok",
+    parsedToolCalls,
+    stopReason,
+    ...anthropicResponse,
+  };
+
+  return response;
+};
+
+const getOpenAiResponse = async (
+  params: OpenAiLlmParams,
+): Promise<LlmResponse<OpenAiLlmParams>> => {
+  const { tools, trimMessageAtIndex, retryCount = 0 } = params;
+
+  const openAiTools = tools?.map(mapLlmToolDefinitionToOpenAiToolDefinition);
+
+  const completionPayload: ChatCompletionCreateParamsNonStreaming = {
+    ...params,
+    tools: openAiTools,
+    stream: false,
+  };
+
+  /**
+   * @todo: consider removing the message trimming functionality,
+   * in favor of forcing the caller to chunk calls to the API
+   * according to the model's context window size.
+   */
+  let estimatedPromptTokens: number | undefined;
+  if (typeof trimMessageAtIndex === "number") {
+    const modelContextWindow = modelToContextWindow[params.model];
+
+    const completionPayloadOverhead = 4_096;
+
+    let excessTokens: number;
+    do {
+      estimatedPromptTokens = promptTokensEstimate({
+        messages: params.messages,
+        functions: openAiTools?.map((tool) => tool.function),
+      });
+
+      logger.info(`Estimated prompt tokens: ${estimatedPromptTokens}`);
+
+      excessTokens =
+        estimatedPromptTokens + completionPayloadOverhead - modelContextWindow;
+
+      if (excessTokens < 10) {
+        break;
+      }
+
+      logger.info(
+        `Estimated prompt tokens (${estimatedPromptTokens}) + completion token overhead (${completionPayloadOverhead}) exceeds model context window (${modelContextWindow}), trimming original user text input by ${
+          excessTokens / 4
+        } characters.`,
+      );
+
+      const firstUserMessageContent =
+        completionPayload.messages[trimMessageAtIndex]!.content;
+
+      completionPayload.messages[trimMessageAtIndex]!.content =
+        firstUserMessageContent?.slice(
+          0,
+          firstUserMessageContent.length - excessTokens * 4,
+        ) ?? "";
+    } while (excessTokens > 9);
+  }
+
+  let openAiResponse: ChatCompletion;
+
+  try {
+    openAiResponse = await openai.chat.completions.create(completionPayload);
+
+    logger.debug(`OpenAI response: ${stringify(openAiResponse)}`);
+  } catch (error) {
+    logger.error(`OpenAI API error: ${stringify(error)}`);
+
+    return {
+      status: "api-error",
+    };
+  }
+
+  const retry = async (retryParams: {
+    retryMessages: OpenAiLlmParams["messages"];
+  }): Promise<LlmResponse<OpenAiLlmParams>> => {
+    if (retryCount > maxRetryCount) {
+      return {
+        status: "exceeded-maximum-retries",
+      };
+    }
+
+    logger.debug(
+      `Retrying OpenAi call with the following retry messages: ${stringify(retryParams.retryMessages)}`,
```

Comment:
```suggestion
      `Retrying OpenAI call with the following retry messages: ${stringify(retryParams.retryMessages)}`,
```

Reply:
Thanks, addressed in https://github.com/hashintel/hash/pull/4318/commits/f97fda48b5855f1725daa645196a3da68373818b